File: scanner/module/http/dsstore.go
package http

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"net/http"

	"github.com/bloom42/phaser/common/phaser"
	"github.com/bloom42/phaser/common/phaser/findings"
	"github.com/bloom42/phaser/scanner/module"
)

type DSStoreFileDisclosure struct{}

func (DSStoreFileDisclosure) Name() string {
	return "http/dsstore_file_disclosure"
}

func (DSStoreFileDisclosure) Description() string {
	return "Check for .DS_Store file disclosure"
}

func (DSStoreFileDisclosure) Author() string {
	return "Sylvain Kerkour <sylvain@kerkour.com>"
}

func (DSStoreFileDisclosure) Version() string {
	return "0.1.0"
}

func (DSStoreFileDisclosure) Run(scan *phaser.Scan, target *phaser.Target, port phaser.Port) (module.Result, []error) {
	errs := []error{}
	var ret module.Result
	protocol := "http"

	if !port.HTTP && !port.HTTPS {
		return ret, errs
	}
	if port.HTTPS {
		protocol = "https"
	}

	URL := fmt.Sprintf("%s://%s:%d/.DS_Store", protocol, target.Host, port.ID)
	req, err := http.NewRequest("GET", URL, nil)
	if err != nil {
		errs = append(errs, err)
		return ret, errs
	}
	req.Header.Set("User-Agent", scan.Profile.HTTP.UserAgent)
	res, err := scan.HTTPClient.Do(req)
	if err != nil {
		errs = append(errs, err)
		return ret, errs
	}
	// response, _ := httputil.DumpResponse(res, true)
	body, err := ioutil.ReadAll(res.Body)
	if err != nil {
		errs = append(errs, err)
		return ret, errs
	}
	res.Body.Close()

	if bytes.Equal(body[0:8], []byte{0x0, 0x0, 0x0, 0x1, 0x42, 0x75, 0x64, 0x31}) {
		ret = findings.URL{URL: URL}
	}

	return ret, errs
}

File: src/scanner/modules/http/ds_store.rs
use crate::{
    scanner::{
        module,
        findings,
        Scan,
        Target,
    },
    error::PhaserError,
};


pub struct DsStore{}

impl module::BaseModule for DsStore {
    fn name(&self) -> String {
        return "http/ds-store".to_string();
    }

    fn description(&self) -> String {
        return "Check for .DS_Store file disclosure".to_string();
    }

    fn author(&self) -> String {
        return "Sylvain Kerkour <sylvain@kerkour.com>".to_string();
    }

    fn version(&self) -> String {
        return "0.1.0".to_string();
    }
}

// TODO: handle error
impl module::PortModule for DsStore {
    fn run(&self, _: &Scan, target: &Target, port: &findings::Port) -> Result<findings::Data, PhaserError> {
        let protocol = if port.http {
            "http"
        } else if port.https {
            "https"
        } else {
            ""
        };

        if protocol == "" {
            return Ok(findings::Data::None);
        }

        let url = format!("{}://{}:{}/.DS_Store", &protocol, &target.host, &port.id);
        let mut body = reqwest::get(&url)?;

        let mut buf: Vec<u8> = vec!();
        body.copy_to(&mut buf)?;


        if is_ds_store(&buf) {
            return Ok(findings::Data::Url(findings::Url{
                url,
            }));
        }

        return Ok(findings::Data::None);
    }
}

fn is_ds_store(content: &[u8]) -> bool {
    if content.len() < 8 {
        return false;
    }

    let signature = [0x0, 0x0, 0x0, 0x1, 0x42, 0x75, 0x64, 0x31];

    return &content[0..8] == &signature;
}


#[cfg(test)]
mod tests {
    use crate::scanner::module::BaseModule;

    #[test]
    fn module_name() {
        let module = super::DsStore{};
        assert_eq!("http/ds-store", module.name());
    }

    #[test]
    fn is_dotenv() {
        let body = "Aswswswsw";
        let body2 = [0x00, 0x00, 0x00, 0x01, 0x42, 0x75, 0x64, 0x31, 0x00, 0x00, 0x30, 0x00, 0x00, 0x00, 0x08, 0x0];

        assert_eq!(false, super::is_ds_store(body.as_bytes()));
        assert_eq!(true, super::is_ds_store(&body2));
    }

}
